webapp/src/components/todo_item/todo_item.jsx

```diff
@@ -0,0 +1,337 @@
+    const [editTodo, setEditTodo] = useState(false);
+    const [message, setMessage] = useState(issue.message);
+    const [description, setDescription] = useState(issue.description);
+    const [hidden, setHidden] = useState(false);
+
+    const style = getStyle(theme);
+
+    const handleClick = (e) => handleFormattedTextClick(e);
+
+    const htmlFormattedMessage = PostUtils.formatText(issue.message, {
+        siteURL,
+    });
+
+    const htmlFormattedDescription = PostUtils.formatText(issue.description, {
+        siteURL,
+    });
+
+    const issueMessage = PostUtils.messageHtmlToComponent(htmlFormattedMessage);
+    const issueDescription = PostUtils.messageHtmlToComponent(htmlFormattedDescription);
+
+    let listPositionMessage = '';
+    if (issue.user) {
+        if (issue.list === '') {
+            listPositionMessage =
+                'Accepted. On position ' + (issue.position + 1) + '.';
+        } else if (issue.list === 'in') {
+            listPositionMessage =
+                'In Inbox on position ' + (issue.position + 1) + '.';
+        } else if (issue.list === 'out') {
+            listPositionMessage = '';
+        }
+    }
+
+    const listDiv = (
+        <div
+            className='light'
+            style={style.subtitle}
+        >
+            {listPositionMessage}
+        </div>
+    );
+
+    const acceptButton = (
+        <AcceptButton
+            issueId={issue.id}
+            accept={accept}
+        />
+    );
+
+    const onKeyDown = (e) => {
+        if (e.key === 'Enter') {
+            saveEditedTodo();
+        }
+
+        if (e.key === 'Escape') {
+            setEditTodo(false);
+        }
+    };
+
+    const actionButtons = (
+        <div className='todo-action-buttons'>
+            {canAccept(list) && acceptButton}
+        </div>
+    );
+
+    let removeTimeout;
+    const completeTimeout = useRef(null);
+
+    const completeToast = useCallback(() => {
+        openTodoToast({icon: 'check', message: 'Todo completed', undo: undoCompleteTodo});
+
+        setHidden(true);
+
+        completeTimeout.current = setTimeout(() => {
+            complete(issue.id);
+        }, 5000);
+    }, [complete, openTodoToast, issue, completeTimeout]);
+
+    const undoRemoveTodo = () => {
+        clearTimeout(removeTimeout);
+        setHidden(false);
+    };
+
+    const undoCompleteTodo = () => {
+        clearTimeout(completeTimeout.current);
+        setHidden(false);
+        setDone(false);
+    };
+
+    const completeButton = (
+        <CompleteButton
+            active={done}
+            theme={theme}
+            issueId={issue.id}
+            markAsDone={() => setDone(true)}
+            completeToast={completeToast}
+        />
+    );
+
+    const removeTodo = useCallback(() => {
+        openTodoToast({icon: 'trash-can-outline', message: 'Todo deleted', undo: undoRemoveTodo});
+        setHidden(true);
+        removeTimeout = setTimeout(() => {
```

Comment:
This won't work. You need to use a reference, like you did with completeTimeout.

The reason is the following:
- Every render, a new reference to removeTimeout is created.
- The value of removeTimeout (undefined) is used to create undoRemoveTodo.
- RemoveTodo sets a value to the removeTimeout variable, but this does not get reflected in the body of `undoRemoveTodo`, since it is already using old value (undefined).

If we use a reference:
- On every render, the reference to removeTimeout is the same (a reference to an object like `{current: something}`)
- undoCompleteTodo receives the value of removeTimeout (a reference to the object `{current: something}`)
- removeTodo assigns a new value not to removeTimeout, but to removeTimeout.current. Since we are working with references to objects here, the new value is ready to be referenced on `undoCompleteTodo`

I hope it is a bit clearer now.